File: interview-questions-cpp/top-interview-questions/MaxDepthTest.cpp
// https://leetcode.com/problems/maximum-depth-of-binary-tree/description/
#include <gtest/gtest.h>
#include <memory>
#include <algorithm>

namespace {
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}

  ~TreeNode() {
    if (left != nullptr){
      delete left;
      left = nullptr;
    }

    if (right != nullptr) {
      delete right;
      right = nullptr;
    }
  }
};


int maxDepth(TreeNode* root) {
  if (root== nullptr) {
    return 0;
  }

  return (1 + std::max(maxDepth(root->left), maxDepth(root->right)));
}

std::unique_ptr<TreeNode> createTreeTC1() {
  auto root = std::make_unique<TreeNode>(3);
  root->left = new TreeNode(9);
  root->right = new TreeNode(20);
  root->right->left = new TreeNode(15);
  root->right->right = new TreeNode(7);
  return root;
}

} // namespace

TEST(MaxDepth, TC1) {
  auto tree = createTreeTC1();
  auto actualDepth = maxDepth(tree.get());
  ASSERT_EQ(3, actualDepth);
}

File: interview-questions-kotlin/src/test/kotlin/dlind1974/algo/top/MaxDepthTest.kt
package dlind1974.algo.top

import kotlin.math.max
import kotlin.test.Test
import kotlin.test.assertEquals

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

private fun maxDepth(root: TreeNode?): Int {
    if (root == null) return 0
    return 1 + max(maxDepth(root.left), maxDepth(root.right))
}

private fun maxDepthIdiomatic(root: TreeNode?): Int =
    root?.let { 1 + max(maxDepthIdiomatic(it.left), maxDepthIdiomatic(it.right)) } ?: 0


private fun createTreeTc1(): TreeNode {
    val root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right!!.left = TreeNode(25)
    root.right!!.right = TreeNode(7)
    return root
}

class MaxDepthBinaryTreeTest {

    @Test
    fun testTc1() {
        val tree = createTreeTc1()
        val actualDepth = maxDepth(tree)
        assertEquals(3, actualDepth)

        val actualDepth2 = maxDepthIdiomatic(tree)
        assertEquals(3, actualDepth2)

    }
}